File: src/client_web/src/lib/business-day.ts
import type { BranchResponse } from '@/api/branch'
import { dayStartHour, isOvernightShift } from './branch'

/**
 * Start of the branch's current business day (mobile parity): overnight
 * shifts (e.g. 17:00 → 05:00) that haven't reached the start hour yet still
 * belong to yesterday's shift.
 */
export function businessDayStart(branch?: BranchResponse | null): Date {
  const startHour = dayStartHour(branch)
  const overnight = isOvernightShift(branch)
  const now = new Date()
  const start = new Date(now)
  if (overnight && now.getHours() < startHour) {
    start.setDate(start.getDate() - 1)
  }
  start.setHours(startHour, 0, 0, 0)
  return start
}

File: src/client_app/lib/core/utils/business_day.dart
import '../models/branch.dart';

/// When the branch's current business day started. An overnight shift
/// (17:00 → 05:00) that has not reached its start hour yet is still
/// yesterday's; a same-day shift always starts today.
DateTime businessDayStart(Branch? branch) {
  final startHour = branch?.dayStartHour ?? 17;
  final overnight = branch?.isOvernightShift ?? true;
  final now = DateTime.now();
  if (overnight && now.hour < startHour) {
    return DateTime(now.year, now.month, now.day - 1, startHour);
  }
  return DateTime(now.year, now.month, now.day, startHour);
}

/// The shift day a moment belongs to, as that day's local midnight: before
/// the start hour of an overnight shift it is the previous day's.
DateTime shiftDayOf(DateTime moment, Branch? branch) {
  final startHour = branch?.dayStartHour ?? 17;
  final overnight = branch?.isOvernightShift ?? true;
  final local = moment.toLocal();
  final day = overnight && local.hour < startHour ? local.day - 1 : local.day;
  return DateTime(local.year, local.month, day);
}
